service/lib/dinstaller/dbus/question.rb

```diff
@@ -0,0 +1,172 @@
+# frozen_string_literal: true
+
+# Copyright (c) [2022] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "dbus"
+require "pathname"
+require "dinstaller/dbus/questions"
+require "dinstaller/question"
+require "dinstaller/luks_question"
+
+module DInstaller
+  module DBus
+    # Class to represent a question on D-Bus
+    #
+    # Questions are dynamically exported on D-Bus. All questions are exported as children of
+    # {DBus::Questions} object to mimic ObjectManager behavior. Note that ruby-dbus does not support
+    # ObjectManager yet.
+    #
+    # Clients should provide an answer for each question.
+    class Question < ::DBus::Object
+      # This module contains the interfaces a question can implement. The interfaces that a question
+      # implements are dynamically determined while creating a new question.
+      module Interfaces
+        # Generic interface for a question
+        module Question
+          QUESTION_INTERFACE = "org.opensuse.DInstaller.Question1"
+          private_constant :QUESTION_INTERFACE
+
+          # Unique id of the question
+          #
+          # @return [Integer]
+          def id
+            backend.id
+          end
+
+          # Text of the question
+          #
+          # @return [String]
+          def text
+            backend.text
+          end
+
+          # Options the question admits as answer
+          #
+          # @note Clients are responsible of generating the proper label for each option.
+          #
+          # @return [Array<String>]
+          def options
+            backend.options.map(&:to_s)
+          end
+
+          # Default option a client should offer as answer
+          #
+          # @return [String]
+          def default_option
+            backend.default_option.to_s
+          end
+
+          # Answer selected for a client
+          #
+          # @return [String]
+          def answer
+            backend.answer.to_s
+          end
+
+          # Selects an option as answer
+          #
+          # @param option [String]
+          def answer=(option)
+            backend.answer = option.to_sym
```

Comment:
I kind of expect here validation if answer is from options before passing to_sym otherwise we can be easy DOSed by passing random strings as answer.

Reply:
Yes, I already had it on mind. I plan to start adding validations meanwhile I write unit tests. But I did prefer to not do it until validating the general approach.